### ISSUE TITLE
Assignment question 4.2

### ISSUE THREAD

**Issue by fjhohura789** [STAT540-UBC/Discussion#324]
According to the example of Paul Pavlidis (https://github.com/STAT540-UBC/STAT540-UBC.github.io/blob/master/examples/rnaseqdiffex-examples/examples-RNAseq.Rmd) (line 240) "voom() takes counts, NOT cpm". Do we need to change our CPM expression data to count? Or we have to use the CPM data?

**Comment by ppavlidis** [comment]
`voom` takes counts ("Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear modelling.")

**Comment by jafarzadeh91** [comment]
Hi,
You can work around the issue by multiplying CPM values by 10^6 before giving the numbers as inputs to the `voom` function.

**Comment by ShuxianFan** [comment]
Then for Q5.4, should we be using the value in the matrix that voom(CPM*10^6,...) return? (the log2 transformed value) or could we log2 transform directly using the CPM expression data? I am assuming they are equivalent?

**Comment by achenderson** [comment]
@jafarzadeh91 This might be why everyone was getting such different answers for 5.1. I changed to x 10^6 and my answers changed a lot? Not sure if that is useful....

**Comment by jafarzadeh91** [comment]
> @jafarzadeh91 This might be why everyone was getting such different answers for 5.1. I changed to x 10^6 and my answers changed a lot? Not sure if that is useful...

Yeah. If we do not scale the values up to the count numbers we may miss some of the significant genes. The reason is interesting to mention here:
Voom adds a value of 0.5 to all the presumably "counts" to prevent the log(0) issue for genes with zero counts.
Now when our inputs are CMPs, adding 0.5 is a big deal and just masks the differences in the expression of genes across samples.

Sorry for the confusing language in this question. We TAs work to refine this question in case we decide to use the question for the students in the next semester. For now, my best advice is to transfer values to their real scale before applying `voom`. As it has been mentioned by Paul too, otherwise the results are not reliable.

**Comment by ShuxianFan** [comment]
But why voomResult$E and log2(CPM+0.5) give me different values? @jafarzadeh91

**Comment by achenderson** [comment]
Yes I am very confused. When I tried to fix my initial code by doing x10^6, I now get zero values that have significant change across cell types. Any obvious errors here?

y <- DGEList(count=expressionDat2*10^6)
keep <- rowSums(cpm(y)>1) >= 4
y1 <- y[keep, , keep.lib.sizes=FALSE]
va<-voom(y1, bigdesign, plot=FALSE)
vfita<-(lmFit(va, bigdesign) %>% eBayes())
summary(decideTests(vfita, p.value=.05, adjust.method="fdr"))
diffgenes<-toptable(vfita, number=2000, coef=3, adjust.method="fdr", p.value=.05)
meanpv<-mean(diffgenes$adj.P.Val)

**Comment by j-schaub** [comment]
@achenderson I also get 0 values (I was getting 3 before I scaled everything back to full counts)

**Comment by achenderson** [comment]
@j-schaub me too!

**Comment by ShuxianFan** [comment]
@jafarzadeh91  For the last question, should we use voomResult$E as the logCPM or should we log2(CPM) using our original data matrix? They give me different values! Thanks!

**Comment by ShuxianFan** [comment]
@achenderson same here. 0 values now and 3 before.

**Comment by achenderson** [comment]
if i use the global method, i get hits but i have no idea what i am doing.

**Comment by achenderson** [comment]
Actually, I get hits if I do "decideTests" but not topTable which is slightly disturbing? The following lines give me 123 and 0 genes differentially expressed, respectively. 

summary(decideTests(vfita, p.value=.05, adjust.method="fdr", method = "global"))
diffgenes<-toptable(vfita, number=2000, coef=3, adjust.method="fdr", method = "global", p.value=.05)

**Comment by j-schaub** [comment]
I didn't specify a method with decideTests and I get 0 significant results (same as topTable). I guess global isn't the default?

**Comment by tfsevers88** [comment]
I’m getting 0 genes for differential expression with respect to cell type now, too. I was getting 1 gene, but realized I was using expression data that I had log transformed in assignment 1. At least now I’m consistent with others...progress?

**Comment by jafarzadeh91** [comment]
> @jafarzadeh91 For the last question, should we use voomResult$E as the logCPM or should we log2(CPM) using our original data matrix? They give me different values! Thanks!

Both log2 and Voom have the same goal: Stabilizing variance! Both have assumptions on the relationship between the mean and variance and one can be better than the other depending on the data at hand.
(See:https://en.wikipedia.org/wiki/Variance-stabilizing_transformation)

It seems Voom is more realistic than log transforms according to their published results, but for this question, it asked you to use logCMP so please use this value in your code.

**Comment by nknoetze** [comment]
I have the same results with having no DE genes!

**Comment by vahidAK** [comment]
No DE genes, too. What should we do

**Comment by jafarzadeh91** [comment]
Hey guys, Because we do not know the total number of counts we cannot recover them simply by multiplication. Please stay tuned! we will figure it out as soon as we could.
